libtsuba/src/RDG.cpp

```diff
@@ -109,6 +116,84 @@ MasterPropName(unsigned i) {
   return std::string(kMasterNodesPropName) + "_" + std::to_string(i);
 }
 
+katana::Result<std::shared_ptr<arrow::ChunkedArray>>
+LargeStringToChunkedString(
+    const std::shared_ptr<arrow::LargeStringArray>& arr) {
+  std::vector<std::shared_ptr<arrow::Array>> chunks;
+
+  arrow::StringBuilder builder;
+
+  uint64_t inserted = 0;
+  constexpr uint64_t kMaxSize = 2147483646;
+
+  for (uint64_t i = 0, size = arr->length(); i < size; ++i) {
+    if (!arr->IsValid(i)) {
+      auto status = builder.AppendNull();
+      if (!status.ok()) {
+        KATANA_LOG_ERROR("could append null: {}", status);
+        return tsuba::ErrorCode::ArrowError;
+      }
+      continue;
+    }
+    arrow::util::string_view val = arr->GetView(i);
+    uint64_t val_size = val.size();
+    KATANA_LOG_ASSERT(val_size < kMaxSize);
+    if (inserted + val_size >= kMaxSize) {
+      std::shared_ptr<arrow::Array> new_arr;
+      auto status = builder.Finish(&new_arr);
+      if (!status.ok()) {
+        KATANA_LOG_ERROR("could finish building string array: {}", status);
+        return tsuba::ErrorCode::ArrowError;
+      }
+      chunks.emplace_back(new_arr);
+      inserted = 0;
+      builder.Reset();
+    }
+    inserted += val_size;
+    auto status = builder.Append(val);
+    if (!status.ok()) {
+      KATANA_LOG_ERROR("could not add string to array builder: {}", status);
+      return tsuba::ErrorCode::ArrowError;
+    }
+  }
+  if (inserted > 0) {
+    std::shared_ptr<arrow::Array> new_arr;
+    auto status = builder.Finish(&new_arr);
+    if (!status.ok()) {
+      KATANA_LOG_ERROR("could finish building string array: {}", status);
+      return tsuba::ErrorCode::ArrowError;
+    }
+    chunks.emplace_back(new_arr);
+  }
+
+  auto maybe_res = arrow::ChunkedArray::Make(chunks, arrow::utf8());
+  if (!maybe_res.ok()) {
+    KATANA_LOG_ERROR(
+        "could not make arrow chunked array: {}", maybe_res.status());
+    return tsuba::ErrorCode::ArrowError;
+  }
+  return maybe_res.ValueOrDie();
+}
+
+// HandleBadParquetTypes here and HandleBadParquetTypes in AddProperties.cpp
+// workaround a libarrow2.0 limitation in reading and writing LargeStrings to
+// parquet files.
+katana::Result<std::shared_ptr<arrow::ChunkedArray>>
+HandleBadParquetTypes(std::shared_ptr<arrow::ChunkedArray> old_array) {
+  if (old_array->num_chunks() > 1) {
+    return old_array;
+  }
+  switch (old_array->type()->id()) {
```

Comment:
I'm curious about this switch statement with one case, does it have an advantage over if/else?

Reply:
Not really; It's mover of a personal style/habit to use switch statements to handle control flow that depends on enums like this.